org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/XtextBuilder.java

```diff
@@ -319,15 +320,62 @@ protected void incrementalBuild(IResourceDelta delta, final IProgressMonitor mon
 		progress.worked(2);
 		Task task = closedProjectsQueue.exhaust();
 		try {
-			addInfosFromTask(task, toBeBuilt);
-			doBuild(toBeBuilt, progress.split(8), BuildType.INCREMENTAL);
+			addInfosFromTaskAndBuild(task, toBeBuilt, BuildType.INCREMENTAL, progress.split(8));
 		} catch(Exception e) {
 			task.reschedule();
 			throw e;
 		}
 	}
 
-	private void addInfosFromTask(Task task, final ToBeBuilt toBeBuilt) {
+	private static Boolean mustCallDeprecatedDoBuild = null;
+	private static boolean wasDeprecationWarningLogged = false;
+	
+	/**
+	 * @since 2.18
+	 */
+	protected void addInfosFromTaskAndBuild(Task task, final ToBeBuilt toBeBuilt, BuildType buildType, IProgressMonitor monitor)
+			throws CoreException {
+		addInfosFromTask(task, toBeBuilt);
+		if (XtextBuilder.class.equals(getClass())) {
+			doBuild(toBeBuilt, task.getProjectNames(), monitor, buildType);	
+		} else {
+			if (mustCallDeprecatedDoBuild == null) {
+				mustCallDeprecatedDoBuild = isDoBuildSpecialized(getClass());
+				if (mustCallDeprecatedDoBuild.booleanValue() && !wasDeprecationWarningLogged) {
+					log.warn("XtextBuilder is specialized and overrides deprecated method doBuild(..). The implementation should be adjusted.");
+					wasDeprecationWarningLogged = true;
+				}
+			}
+			if (mustCallDeprecatedDoBuild.booleanValue()) {
+				doBuild(toBeBuilt, monitor, buildType);
+			} else {
+				doBuild(toBeBuilt, task.getProjectNames(), monitor, buildType);
+			}
+		}
+		
+	}
+
+	/**
+	 * @since 2.18
+	 */
+	protected boolean isDoBuildSpecialized(Class<?> c) {
```

Comment:
i wonder if we can do something not to keep this stuff forever e.g. announcing a removal in 2020

Reply:
I'll mark it as deprecated immediately.

Reply:
(same for the others that are for compatibility)

Reply:
I think I caught the relevant candidates.